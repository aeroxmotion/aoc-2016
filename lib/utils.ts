export async function readInput(day: number): Promise<string> {
  if (!Bun.env.SESSION) {
    throw new Error("Missing env session");
  }

  const response = await fetch(
    `https://adventofcode.com/2016/day/${day}/input`,
    {
      headers: {
        Cookie: `session=${Bun.env.SESSION}`,
      },
    },
  );

  return (await response.text()).trim();
}
