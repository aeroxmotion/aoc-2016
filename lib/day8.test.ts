import { test, expect } from "bun:test";

import { readInput } from "./utils";

let input = await readInput(8);

enum Light {
  On = "#",
  Off = ".",
}

function solution<T extends boolean>(
  code: T,
): T extends true ? string : number {
  const WIDTH = 50;
  const HEIGHT = 6;

  let lightsOn = 0;
  const screen: Light[][] = Array.from({ length: HEIGHT }, () =>
    Array(WIDTH).fill(Light.Off),
  );

  for (const line of input.split("\n")) {
    const [ins, ...args] = line.split(" ");

    if (ins === "rect") {
      const [width, height] = args[0].split("x").map(Number);

      for (let i = 0; i < height; i++) {
        for (let j = 0; j < width; j++) {
          if (screen[i][j] === Light.Off) {
            screen[i][j] = Light.On;
            lightsOn++;
          }
        }
      }
    } else {
      const [_, s, __, n] = args;
      const [axis, index] = s.split("=");
      const isRow = axis === "y";

      const idx = +index;
      const R = +n;
      const I = isRow ? WIDTH : HEIGHT;
      const buf: Light[] = Array(I);

      // Rotate right
      for (let i = R; i < I + R; i++) {
        const j = i % I;
        const k = i - R;

        const l = isRow ? idx : j;
        const m = isRow ? j : idx;

        buf[j] = screen[l][m];
        screen[l][m] = buf[k] || screen[isRow ? idx : k][isRow ? k : idx];
      }
    }
  }

  if (!code) {
    return lightsOn as any;
  }

  return screen.map((row) => row.join("")).join("\n") as any;
}

test("part1", () => {
  expect(solution(false)).toBe(110);
});

test("part2", () => {
  expect(solution(true)).toMatchSnapshot();
});
