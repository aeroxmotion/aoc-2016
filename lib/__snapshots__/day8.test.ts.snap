// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`part2 1`] = `
"####...##.#..#.###..#..#..##..###..#....#...#..##.
...#....#.#..#.#..#.#.#..#..#.#..#.#....#...#...#.
..#.....#.####.#..#.##...#....#..#.#.....#.#....#.
.#......#.#..#.###..#.#..#....###..#......#.....#.
#....#..#.#..#.#.#..#.#..#..#.#....#......#..#..#.
####..##..#..#.#..#.#..#..##..#....####...#...##.."
`;
